src/cpp/disassembler/disassembler.cpp

```
@@ -0,0 +1,130 @@
+        if (section->get_type() != ELFIO::SHT_PROGBITS) {
+            continue;
+        }
+        print_section_info(section);
+        if (is_text_section(section_name)) {
+            process_text_section(section, state);
+        } else if (is_data_section(section_name)) {
+            process_data_section(section, state);
+            state->reset();
+        } else {
+            std::cout << "Section not recognized as text or data: " << section_name << std::endl;
+        }
+    }
+}
+
+void asm_disassembler::print_section_info(const ELFIO::section* section) {
+    std::string flags;
+    if (section->get_flags() & ELFIO::SHF_ALLOC) flags += "a";
+    if (section->get_flags() & ELFIO::SHF_WRITE) flags += "w";
+    if (section->get_flags() & ELFIO::SHF_EXECINSTR) flags += "x";
+    asm_write.write_directive("");
+    if (is_data_section(section->get_name())) {
+        asm_write.write_directive(".align " + std::to_string(section->get_addr_align()));
+    }
+}
+
+void asm_disassembler::process_text_section(const ELFIO::section* section, std::shared_ptr<disassembler_state> state) {
+    const char* section_data = section->get_data();
+    size_t section_size = section->get_size();
+    for (size_t offset = ELF_SECTION_HEADER_PADDING; offset < section_size;) {
+        uint8_t opcode = *reinterpret_cast<const uint8_t*>(section_data + offset);
+        auto op_it = isa_op_map->find(opcode);
+        if (op_it != isa_op_map->end()) {
+            if (opcode == ALIGN_OPCODE) {
+                state->increment_address(1);
+                ++offset;
+                continue;
+            }
+            auto deserializer = op_it->second.create_deserializer();
+            size_t consumed = deserializer->deserialize(asm_write, state, section_data + offset);
+            offset += consumed;
+        } else {
+            std::cerr << "[DEBUG] Unknown opcode " << static_cast<int>(opcode) << " at position " << offset << "\n";
```

Please throw an `invalid_asm` exception if the opcode is not recognized. Also, you can then use early-exit control flow style.